hls-test-utils/src/Test/Hls.hs

```diff
@@ -9,38 +10,117 @@ module Test.Hls
+-- | Run 'defaultMainWithRerun', limiting each single test case running at most 10 minutes
 defaultTestRunner :: TestTree -> IO ()
-defaultTestRunner =
-  defaultMainWithIngredients
-    [antXMLRunner, rerunningTests [listingTests, consoleTestReporter]]
+defaultTestRunner = defaultMainWithRerun . adjustOption (const $ mkTimeout 600000000)
 
 gitDiff :: FilePath -> FilePath -> [String]
 gitDiff fRef fNew = ["git", "diff", "--no-index", "--text", "--exit-code", fRef, fNew]
 
 goldenGitDiff :: TestName -> FilePath -> IO ByteString -> TestTree
 goldenGitDiff name = goldenVsStringDiff name gitDiff
 
-testCommand :: String
-testCommand = "test-server"
+runSessionWithServer :: PluginDescriptor IdeState -> FilePath -> Session a -> IO a
+runSessionWithServer plugin = runSessionWithServer' [plugin] def def fullCaps
+
+runSessionWithServerFormatter :: PluginDescriptor IdeState -> String -> FilePath -> Session a -> IO a
+runSessionWithServerFormatter plugin formatter =
+  runSessionWithServer'
+    [plugin]
+    def {formattingProvider = T.pack formatter}
+    def
+    fullCaps
+
+-- | Run an action, with stderr silenced
+silenceStderr :: IO a -> IO a
+silenceStderr action = withTempFile $ \temp ->
+  bracket (openFile temp ReadWriteMode) hClose $ \h -> do
+    old <- hDuplicate stderr
+    buf <- hGetBuffering stderr
+    h `hDuplicateTo'` stderr
+    action `finally` do
+      old `hDuplicateTo'` stderr
+      hSetBuffering stderr buf
+      hClose old
+
+-- | Host a server, and run a test session on it
+-- Note: cwd will be shifted into @root@ in @Session a@
+runSessionWithServer' ::
+  -- | plugins to load on the server
+  [PluginDescriptor IdeState] ->
+  -- | lsp config for the server
+  Config ->
+  -- | config for the test session
+  SessionConfig ->
+  ClientCapabilities ->
+  FilePath ->
+  Session a ->
+  IO a
+runSessionWithServer' plugin conf sconf caps root s = silenceStderr $ do
+  (inR, inW) <- createPipe
+  (outR, outW) <- createPipe
+  -- restore cwd after running the session; otherwise the path to test data will be invalid
+  cwd <- getCurrentDirectory
+  server <-
+    async $
+      Ghcide.defaultMain
+        def
+          { argsHandleIn = pure inR,
+            argsHandleOut = pure outW,
+            argsDefaultHlsConfig = conf,
+            argsLogger = pure noLogging,
+            argsIdeOptions = \config sessionLoader ->
+              let ideOptions = (argsIdeOptions def config sessionLoader) {optTesting = IdeTesting True}
+               in ideOptions {optShakeOptions = (optShakeOptions ideOptions) {shakeThreads = 2}},
+            argsHlsPlugins = pluginDescToIdePlugins $ plugin ++ Ghcide.descriptors
+          }
+
+  x <-
+    runSessionWithHandles inW outR sconf caps root s
+      `finally` setCurrentDirectory cwd
+  timeout 3 (wait server) >>= \case
+    Just () -> pure ()
+    Nothing -> putStrLn "Server does not exit on time, canceling the async task..." >> cancel server
```

Comment:
Tests on Windows passed with this. We should investigate why the server didn't exit properly.

Reply:
I'm just looking at the exit code in ghcide and I think it might be time to update it. I'm not sure that the `waitAnyCancel` construction is needed any more, and it could be hiding exceptions coming from the `lsp` library:

https://github.com/haskell/haskell-language-server/blob/a4dee2d078dc85fb450ec9b35851d3ba3d84789d/ghcide/src/Development/IDE/LSP/LanguageServer.hs#L114-L120

Reply:
Another problem is that `defaultMain` does not call [`setupLogger`](https://hackage.haskell.org/package/lsp-1.2.0.0/docs/Language-LSP-Server.html#v:setupLogger) so we might be missing lsp errors

Reply:
https://github.com/haskell/haskell-language-server/blob/6d1f1a55e39a9fd905bea13d6144125fcb42a594/ghcide/src/Development/IDE/LSP/LanguageServer.hs#L59-L62

These comments match the old code:

https://github.com/haskell/haskell-language-server/blob/fa671cbe1ee8f161df8d79d7a23962239d4d9e75/ghcide/src/Development/IDE/LSP/Server.hs#L73-L99

but now this barrier will be signalled when the server receives exit message instead.

I think we could try removing the `waitAnyCancel`.

Reply:
Yes, now we signal the barrier from two places - the exit handler and also if the initialisation fails, which probably explains the error that you were seeing earlier.

I'm not convinced that removing it will actually fix anything. `runServerWithHandles` doesn't actually exit until the input stream is closed, and I don't know whether `lsp-test` closes the stream until the server exits

Reply:
It seems that `lsp-test` won't close the two handles for us (in our case `mServerProc` is `Nothing`):
https://github.com/haskell/haskell-language-server/blob/7a2ff3ad87b75812880bd580a98b120615bd1ef4/lsp-test/src/Language/LSP/Test/Session.hs#L282-L294

But I didn't find any manual close in the test of `lsp-test`. Maybe @wz1000 could help?

Reply:
I'm pretty sure the handle closes when the server exits.

Reply:
The handle closes when the server exits, sure, but as I said `runServerWithHandles` doesn't exit until the stream ends, so it won't exit.